Emulating HW vrnd

fp64 opened the pull request:
The vrnd stuff from https://github.com/hrydgard/ppsspp/issues/16946 (see https://github.com/hrydgard/ppsspp/issues/16946#issuecomment-1467261209 onwards).

Still needs more testing, just putting it there to get overall comments.
Only tested in couple of games (including Ridge Racer, which is known to use vrnd). Seems to work fine, with no visual differences.
Bumps save state version for `MIPSState`, so save states are probably incompatible.

hrydgard commented on Core/MIPS/MIPS.cpp:
```diff
@@ -308,8 +306,6 @@ void MIPSState::DoState(PointerWrap &p) {
 		Do(p, fcr0_unused);
 	}
 	Do(p, fcr31);
-	Do(p, rng.m_w);
-	Do(p, rng.m_z);
```
Here, you need to check if the savestate version is 3 or below, in which case you still do the old steps, for example to local dummy variables. There are many examples around the code for this.

In this case it would simply be, I think:

```
if (s <= 2) {
   uint32_t dummy;
   Do(p, dummy);
   Do(p, dummy);
}

With that, old savestates should load fine (though new savestates won't load in old versions, but that's fine.)

}
```

fp64 replied:
You mean `if (s <= 3)`, not `if (s <= 2)`?

hrydgard replied:
Oh sorry yes, you are increasing from 3 to 4, not from 2 to 3.

hrydgard commented on Core/MIPS/MIPSVFPUUtils.cpp:
```diff
@@ -769,6 +769,85 @@ float vfpu_dot(const float a[4], const float b[4]) {
+
+// Redundant currently, since void MIPSState::Init() already
+// does this on its own, but left as-is to be self-contained.
+void vrnd_init_default(uint32_t *rcx)
+{
+	rcx[0] = 0x00000001;
```
for new code we use this bracing style:

```
void vrnd_init_default(uint32_t *rcx) {
}
```

hrydgard commented on Core/MIPS/MIPSVFPUUtils.cpp:
```diff
@@ -769,6 +769,85 @@ float vfpu_dot(const float a[4], const float b[4]) {
+{
+	// The actual RNG state appears to be 5 parts
+	// (32-bit each) stored into the registers as follows:
+	uint32_t A=(rcx[0] & 0xFFFFu) | (rcx[4] << 16);
+	uint32_t B=(rcx[1] & 0xFFFFu) | (rcx[5] << 16);
+	uint32_t C=(rcx[2] & 0xFFFFu) | (rcx[6] << 16);
+	uint32_t D=(rcx[3] & 0xFFFFu) | (rcx[7] << 16);
+	uint32_t E=(((rcx[0] >> 16) & 0xF) <<  0)|
+	           (((rcx[1] >> 16) & 0xF) <<  4)|
+	           (((rcx[2] >> 16) & 0xF) <<  8)|
+	           (((rcx[3] >> 16) & 0xF) << 12)|
+	           (((rcx[4] >> 16) & 0xF) << 16)|
+	           (((rcx[5] >> 16) & 0xF) << 20)|
+	           (((rcx[6] >> 16) & 0xF) << 24)|
+	           (((rcx[7] >> 16) & 0xF) << 28);
```
space around operators, including around = and |

hrydgard reviewed (changes requested):
This is great! Really only have a couple of code style tweaks to suggest, and a savestate compatibility fix.

I suppose you could also file an issue to https://github.com/jpcsp/jpcsp/issues and tell them that you've cracked it :)

fp64 commented:
Fixed, I think.

hrydgard commented:
Are you working in the future btw? :)

![image](https://github.com/hrydgard/ppsspp/assets/130929/d4e01e4e-8ebe-4fc0-868d-0251fda8d5c9)

Think you've got your timezone set wrong.

fp64 committed:
Make vrnd match HW closer

See investigation starting
https://github.com/hrydgard/ppsspp/issues/16946#issuecomment-1467261209
for more details.
Still needs more testing.

fp64 committed:
Fix a bug, and bump savestate version

fp64 committed:
Address feedback
